Fixes #1052 - Lack of altitude and accuracy when we use location

xsteelej opened the pull request:
Removed the code that got rid of Altitude and Accuracy from the location data point.

Closes #1052 

#### What has been done to verify that this works as intended?

1. Configured Google Drive API & Google Sheets API's.
2. Created a blank "results" sheet.
3. Configured ODK Collect to use Google drive.
4. Configured ODK Collect to have a "Fallback submission URL" with the URL of the "results" sheet.
5. Uploaded the Demo form on the ticket (Had to rename it to loc.xml and put <?xml version="1.0"?> at the top of the file).
6. Created and finalised a Geo form, submitting it to the "results" sheet.
7. Opened the "results" sheet to confirm that Altitude and Accuracy are now present.
8. Ran the unit tests to ensure that all tests pass.

**Example output now looks like:**
```
data-meta-instanceID | data-Point | data-Accuracy | data-Altitude
uuid:ef212cee-12ec-4db5-a930-8fc08c96d8db | 53.47558925 -2.56203411 | 15.008 | 70.23913574

85.76
```

#### Why is this the best possible solution? Were any other approaches considered?

The implemented solution means that sending forms to existing sheets will work as they currently do. 
If the sheet is updated to contain <_fieldname_>-altitude and/or <_fieldname_>-accuracy, then the respective fields are filled in.

If the sheet is empty, then the presence of a geo location field will create the columns <_fieldname_>-altitude and <_fieldname_>-accuracy.

#### Are there any risks to merging this code? If so, what are they?

None

#### Do we need any specific form for testing your changes? If so, please attach one.

See the form on the issue #1052 

#### Does this change require updates to documentation? If so, please file an issue at https://github.com/opendatakit/docs/issues/new and include the link below.

xsteelej committed:
Update README.md

xsteelej committed:
Fixes #1052

Removed the code that got rid of Altitude and Accuracy from location data point.

lognaturel commented:
Unfortunately it's not quite so simple! We do want to maintain the column with lat long so that points can be easily mapped. So this is a matter of adding two new columns named `<field-name>-altitude` and `<field-name>-accuracy`. To make matters yet more complicated, we want to make sure existing forms continue working. That is, if the altitude and accuracy columns don't exist, then they either need to be added for subsequent submissions or perhaps ignored altogether. Currently, if any columns are missing, the submission fails.

xsteelej commented:
Ok, Sure - that makes sense now.

xsteelej commented:
Hi @lognaturel, @yanokwa,

I've modified the code so that the following is output to the Sheet:
```
data-meta-instanceID | data-Point | data-Accuracy | data-Altitude
uuid:fec1feed-252f-434b-b886-4803904b5624 | 53.47564474 -2.5621478 | 6.432 | 85.28198242
```

Is that what the expected solution is?

xsteelej committed:
Updated to add columns for Accuracy and Altitude.

xsteelej committed:
Removed the old code that removed Accuracy and Altitude, it has been re-located.

yanokwa commented:
I haven't tried the PR, but the output you posted says you are on the right track! Only thing I'm not 100% sure about is if the headers on master currently include "data-" and if the point has no comma separating lat/lon, but it's been a while since I looked at that code.

xsteelej committed:
Fixes #1052

Updated to add Altitude / Accuracy to the Sheet columns.

lognaturel commented:
The tricky case is when a spreadsheet of results already exists. To try that out, use Collect before the changes, clear out the results sheet and send in a few results. Then update to this branch and try sending in more results. It will fail saying that the columns are missing. As a solution, I propose checking if the missing columns end in `-accuracy` or `-altitude` AND have a prefix that matches an existing column (this is to make sure it's a column added by Collect and not one the user created). If so, remove them from the values to submit. In other words, if we have a legacy results sheet, just keep ignoring the accuracy and altitude and only add them if we have a new results sheet. I think this should probably happen in `disallowMissingColumns`. If someone wants to update their existing results sheet they can manually add the fields.

The fields should be named after the base field. For the sample form you're using, the node that has location information is `/data/Point`. When it gets converted to a gsheets header, it ends up as `data-Point`. The corresponding accuracy and altitude columns should be `data-Point-accuracy` and `data-Point-altitude` (note the lowercase). This makes it possible to have multiple locations per form with accuracy and altitude.

Hope all this makes sense, let me know if you have any questions!

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/tasks/InstanceGoogleSheetsUploader.java:
```diff
@@ -263,16 +267,37 @@ private void insertRow(TreeElement element, String parentKey, String key, File i
+     * @param answers Map of answers
+     * @return column titles
+     */
+    private List<Object> addExtraTitles(@NonNull List<Object> columnTitles, @NonNull Map<String, String> answers) {
```
Another approach that might be more targeted and match the existing approach of getting column titles from the form rather than the results would be to modify `getColumnTitles` with something like the following in the loop over children after adding the child:
```java
if (child.getDataType() == org.javarosa.core.model.Constants.DATATYPE_GEOPOINT) {
    columnTitles.add(getElementTitle(child) + "-accuracy");
    columnTitles.add(getElementTitle(child) + "-altitude";
}

xsteelej replied:
Thanks @lognaturel, that all makes a lot of sense. I'll let you know if I have further questions.

xsteelej committed:
Fixes 1052

The fix now handles two scenarios:

1) Blank / Empty sheet

If a Geo Location field exists, then altitude / accuracy columns will be added to the sheet. The column titles will be the title of the
data field appended with "-accuracy" and "-altitude".

2) Existing Sheet

The "-altitude" and "-accuracy" columns will not automatically be added, by default existing functionality will be maintained.
However, if the columns are manually inserted into the sheet with the title of the Geo Location + "-accuracy" and "-altitude"
then those data items will be exported to the columns.

xsteelej committed:
Lint issues fixed

xsteelej committed:
Modified comments above addAltitudeAndAccuracyTitles to improve the readability.

xsteelej committed:
Merge remote-tracking branch 'upstream/master'

# Conflicts:
#	README.md

xsteelej committed:
Merge remote-tracking branch 'upstream/master'

xsteelej commented:
Hi @lognaturel, @yanokwa,

Looks like the bot didn't add a "needs review" label to this PR?

dcbriccetti commented on collect_app/src/main/java/org/odk/collect/android/tasks/InstanceGoogleSheetsUploader.java:
```diff
@@ -273,6 +283,27 @@ private void insertRow(TreeElement element, String parentKey, String key, File i
+        for (Object sheetTitle : sheetHeaders) {
+            String sheetTitleStr = (String) sheetTitle;
+            if (sheetTitleStr.endsWith(altitudeTitlePostfix) || sheetTitleStr.endsWith(accuracyTitlePostfix)) {
+                if (! columnTitles.contains(sheetTitleStr)) {
```
Hey, someone else who likes a space after the 'not' operator!

xsteelej replied:
:) .. I'll blame it on a typo

xsteelej committed:
Fix #1052

Removed space after "not".

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/tasks/InstanceGoogleSheetsUploader.java:
```diff
@@ -391,10 +428,47 @@ private void createSheetsIfNeeded(TreeElement element) throws UploadException {
-            columnTitles.add(getElementTitle(child));
+            final String elementTitle = getElementTitle(child);
+            columnTitles.add(elementTitle);
+            if (newSheet && child.getDataType() == org.javarosa.core.model.Constants.DATATYPE_GEOPOINT) {
```
why do we add those two columns only if it's a new sheet? is it for backward compatibility?

xsteelej replied:
Yes, to allow existing sheets to work in exactly the same way.

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/tasks/InstanceGoogleSheetsUploader.java:
```diff
@@ -86,6 +87,9 @@
     private final GoogleAccountsManager accountsManager;
 
+    private static final String altitudeTitlePostfix = "-altitude";
+    private static final String accuracyTitlePostfix = "-accuracy";
```
In the whole code, we use capital letters and `_` between words for static final fields so I think it should be good to keep it up.

dcbriccetti replied:
Agreed. For more: https://docs.oracle.com/javase/tutorial/java/nutsandbolts/variables.html

xsteelej replied:
Also Agreed, fix will be imminent.

xsteelej committed:
Updated constant Strings altitudeTitlePostfix and accuracyTitlePostfix to be standardised constant names: ALTITUDE_TITLE_POSTFIX and ACCURACY_TITLE_POSTFIX.

dcbriccetti commented:
It may be time for some commit cleanup. Do you know about
`git rebase -i HEAD~n` ?
https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History

xsteelej commented:
Sure ok, I'll rebase the last two commits

xsteelej commented:
Thanks for the rebase tip @dcbriccetti, it's something that I've very rarely done in the past.

I have rebased back 1 commit to remove the extra commit that I did to force the rebuild. Lo and behold, when I pushed again the lint checks failed, I used the `git commit --amend` trick and pushed again to force a rebuild (Luckily the checks all passed that time!).

Should be good for a review again.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/tasks/InstanceGoogleSheetsUploader.java:
```diff
@@ -391,10 +428,47 @@ private void createSheetsIfNeeded(TreeElement element) throws UploadException {
+     * @param elementTitle - Current header title
+     * @param answer - String that holds the Lat/Long/Altitude/Accuracy
+     */
+    private void parseAltitudeAndAccuracy(@NonNull Map<String, String> answers, @NonNull List<Object> columnTitles, @NonNull String elementTitle, @NonNull String answer) {
```
I think a better name would be `parseGeopoint` since this is also separating out the lat/long. How about not passing in `answers` and returning a new map from column title to value that gets added to `answers` by the caller (`getAnswers`)? I know several other methods in this class mutate their parameters but it makes the code harder to read and in this case I think it's easy to avoid.

xsteelej replied:
Agreed - parseGeopoint is a better name - Renamed.

Passing back a Map is cleaner, i'll do a refactor - I should be able to use something like answers.putAll(parseGeopoint(...));

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/tasks/InstanceGoogleSheetsUploader.java:
```diff
@@ -391,10 +428,47 @@ private void createSheetsIfNeeded(TreeElement element) throws UploadException {
+     * @param answer - String that holds the Lat/Long/Altitude/Accuracy
+     */
+    private void parseAltitudeAndAccuracy(@NonNull Map<String, String> answers, @NonNull List<Object> columnTitles, @NonNull String elementTitle, @NonNull String answer) {
+
```
Extra line break!

xsteelej replied:
Removed

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/tasks/InstanceGoogleSheetsUploader.java:
```diff
@@ -273,6 +283,27 @@ private void insertRow(TreeElement element, String parentKey, String key, File i
     }
 
+    /**
+     * Adds titles ending with "-altitude" or "-accuracy" that may have been added to the
```
This is for the case where a user manually adds the altitude and accuracy columns, right? It took me a while to get that and I think maybe just adding the word "manually" to "may have been manually added to the..." would help with that.

xsteelej replied:
That makes more sense, I'll change to "Adds titles ending with "-altitude" or "-accuracy" that may have been manually added to the ..."

lognaturel reviewed (commented):
Apologies for the delay in getting back to you on this @xsteelej -- I was out last week. Just a couple of small suggestions and then it should be ready for QA!

xsteelej commented:
Hi @lognaturel, hope you had a good break! Thanks for the comments, i'll address those comments shortly.

dcbriccetti commented on collect_app/src/main/java/org/odk/collect/android/tasks/InstanceGoogleSheetsUploader.java:
```diff
@@ -391,10 +428,47 @@ private void createSheetsIfNeeded(TreeElement element) throws UploadException {
 
-    private List<Object> getColumnTitles(TreeElement element) {
+    /**
+     * Strip the Altitude and Accuracy from a location String and add them as separate column if
```
Strip->strips; add->adds
Doc comments verbs use declarative voice, not imperative 
Python does the opposite! :-)

xsteelej replied:
Updated :-)

jrsteele09 committed:
Fix #1052

As per review comments:

- Renamed parseAltitudeAndAccuracy to parseGeopoint
- parseGeopoint returns a map rather than mutating a map parameter
- Updated method comments as recommended by reviewers

jrsteele09 committed:
Merge branch 'master' of https://github.com/xsteelej/collect into issue_#1052_lack_of_altitude_accuracy

jrsteele09 committed:
Merge remote-tracking branch 'upstream/master'

jrsteele09 committed:
Merge branch 'master' into issue_#1052_lack_of_altitude_accuracy

mmarciniak90 commented:
@xsteelej, @lognaturel 
I'm in the middle of testing and I noticed that latitude and longitude are not separated by a comma.
Is it an intended change? 
![1](https://user-images.githubusercontent.com/3907554/39815580-ed9a1002-5387-11e8-8a93-6744f92e938c.png)

xsteelej commented:
Hi @mmarciniak90, thanks for testing. The data-Point field is the same as it was before this change was made. Lat / Long were just separated by a space and not a comma.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/tasks/InstanceGoogleSheetsUploader.java:
```diff
@@ -504,17 +583,6 @@ private String getElementTitle(AbstractTreeElement element) {
-                        int firstSpace = answer.indexOf(' ');
-                        int secondSpace = answer.indexOf(" ", firstSpace + 1);
-                        answer = answer.substring(0, secondSpace);
-                        answer = answer.replace(' ', ',');
```
Good catch, @mmarciniak90. This replacement was lost and should be brought back, @xsteelej.

xsteelej replied:
Yes very good catch @mmarciniak90 & @lognaturel. My mistake - I looked back at what I thought was a sheet that I had generated before my changes, should have looked at the code in Master instead - Apologies!

jrsteele09 committed:
Fix #1052

This update reinstates the code that puts a comma ',' between the Latitude and Longitude fields. This error was spotted during testing.

mmarciniak90 commented:
**Tested with success.**

![11](https://user-images.githubusercontent.com/3907554/39868031-56a5d1fc-5447-11e8-872e-3002cb209cef.png)

Verified on Android: 4.1, 4.2, 4.4, 5.1, 6.0, 7.0, 8.1
Verified cases:
- sent a submission to cleared Google Sheet - all columns are created
- sent an empty submission to cleared Google Sheet - all columns are created
- sent a submission to existed Google Sheet - accuracy and altitude are not collected
- sent a submission after adding columns manually

@opendatakit-bot unlabel "needs testing"
@opendatakit-bot label "behavior verified"

lognaturel commented:
Awesome, thanks, @xsteelej!
